### PR TITLE
For pilot2 d

### PR DESCRIPTION
**Pull request by lauraser**
The following check takes a canvas as input, loops over all figures containing IROCs and OROCs in the canvas and checks for empty pads. 
Besides storing the main quality object, it has private member to store qualities for every histogram which are later used in beautify to show specific quality for every histogram instead of the entire canvas.
@wiechula Stefan suggested the limits for medium quality 30% of pads empty and for bad 60% of pads empty. Do you think this is reasonable?

### CODE REVIEW

**Comment by Barthelemy** [comment]
I let @wiechula  review this PR.

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      std::string titleh = h->GetTitle();
+      bool iroc = false;
+      bool oroc = false;
+      //check if we are dealing with IROC or OROC
+      if (titleh.find("IROC") != std::string::npos)
+        iroc = true;
+      if (titleh.find("OROC") != std::string::npos)
+        oroc = true;
+      if (!(iroc || oroc))
+        return Quality::Null;
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:57]: add readability braces and simplify expression (assuming it cannot be both, IROC and OROC). Also, the IROC / OROC distinction is only used to get the total number of pads.

```suggestion
      int totalPads = 0;
      if (titleh.find("IROC") != std::string::npos) {
        totalPads = 5280;
      } else if (titleh.find("OROC") != std::string::npos) {
        totalPads = 9280;
      } else {
        return Quality::Null;
      }
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+  auto mo = moMap->begin()->second;
+  Quality result = Quality::Null;
+  if (mo->getName() == "c_ROCs_Pedestal_2D") {
+    result = Quality::Good;
+    auto* canv = dynamic_cast<TCanvas*>(mo->getObject());
+    // Check all histograms in the canvas
+    for (int tpads = 1; tpads <= 72; tpads++) {
+      auto padName = fmt::format("c_ROCs_Pedestal_2D_{:d}", tpads);
+      auto histName = fmt::format("h_Pedestals_ROC_{:02d}", tpads - 1);
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:45]:
```suggestion
      const auto padName = fmt::format("c_ROCs_Pedestal_2D_{:d}", tpads);
      const auto histName = fmt::format("h_Pedestals_ROC_{:02d}", tpads - 1);
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+void PadCalibrationCheck2D::configure(std::string) {}
+Quality PadCalibrationCheck2D::check(std::map<std::string, std::shared_ptr<MonitorObject>>* moMap)
+{
+
+  auto mo = moMap->begin()->second;
+  Quality result = Quality::Null;
+  if (mo->getName() == "c_ROCs_Pedestal_2D") {
+    result = Quality::Good;
+    auto* canv = dynamic_cast<TCanvas*>(mo->getObject());
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:41]: dynamic cast only makes sense if you want the verify the object type. So I guess you should either change this to `static_cast` of c-style cast `(TCanvas*)` or add
```c++
if (!canv) {
  return Quality::Null;
}

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    // Check all histograms in the canvas
+    for (int tpads = 1; tpads <= 72; tpads++) {
+      auto padName = fmt::format("c_ROCs_Pedestal_2D_{:d}", tpads);
+      auto histName = fmt::format("h_Pedestals_ROC_{:02d}", tpads - 1);
+      TPad* pad = (TPad*)canv->GetListOfPrimitives()->FindObject(padName.data());
+      TH2F* h = (TH2F*)pad->GetListOfPrimitives()->FindObject(histName.data());
+      std::string titleh = h->GetTitle();
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:48]:
```suggestion
      const std::string titleh = h->GetTitle();
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        return Quality::Null;
+
+      int NX = h->GetNbinsX();
+      int NY = h->GetNbinsY();
+
+      int totalPads = 0;
+      if (iroc)
+        totalPads = 5280;
+      if (oroc)
+        totalPads = 9280;
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:66]: Not needed, see above.
```suggestion
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+      //check if we are dealing with IROC or OROC
+      if (titleh.find("IROC") != std::string::npos)
+        iroc = true;
+      if (titleh.find("OROC") != std::string::npos)
+        oroc = true;
+      if (!(iroc || oroc))
+        return Quality::Null;
+
+      int NX = h->GetNbinsX();
+      int NY = h->GetNbinsY();
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:60]:
```suggestion
      const int NX = h->GetNbinsX();
      const int NY = h->GetNbinsY();
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+          int val = h->GetBinContent(i, j);
+          if (val > 0) {
+            sum += 1;
+          }
+        }
+      }
+
+      // Check how many are off
+      float mediumLimit = 0.7;
+      float badLimit = 0.4;
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:81]: Shouldn't this be configurable from the json file?
```suggestion
      const float mediumLimit = 0.7;
      const float badLimit = 0.4;
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+      if (sum > badLimit * totalPads && sum < mediumLimit * totalPads) {
+        if (result == Quality::Good)
+          result = Quality::Medium;
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:84]:
```suggestion
        if (result == Quality::Good) {
          result = Quality::Medium;
        }
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    auto* canv = dynamic_cast<TCanvas*>(mo->getObject());
+    // Check all histograms in the canvas
+    for (int tpads = 1; tpads <= 72; tpads++) {
+      auto padName = fmt::format("c_ROCs_Pedestal_2D_{:d}", tpads);
+      auto histName = fmt::format("h_Pedestals_ROC_{:02d}", tpads - 1);
+      TPad* pad = (TPad*)canv->GetListOfPrimitives()->FindObject(padName.data());
+      TH2F* h = (TH2F*)pad->GetListOfPrimitives()->FindObject(histName.data());
+      std::string titleh = h->GetTitle();
+      bool iroc = false;
+      bool oroc = false;
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:50]: Not needed, see below.
```suggestion
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+  }
+  return result;
+}
+
+std::string PadCalibrationCheck2D::getAcceptedType() { return "TCanvas"; }
+
+void PadCalibrationCheck2D::beautify(std::shared_ptr<MonitorObject> mo, Quality)
+{
+
+  auto* canv = dynamic_cast<TCanvas*>(mo->getObject());
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:105]: See comment above.

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+  // Loop over all histograms and beautify accordingly to specific quality object
+  for (int tpads = 1; tpads <= 72; tpads++) {
+    auto padName = fmt::format("c_ROCs_Pedestal_2D_{:d}", tpads);
+    auto histName = fmt::format("h_Pedestals_ROC_{:02d}", tpads - 1);
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:109]:
```suggestion
    const auto padName = fmt::format("c_ROCs_Pedestal_2D_{:d}", tpads);
    const auto histName = fmt::format("h_Pedestals_ROC_{:02d}", tpads - 1);
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    TPad* pad = (TPad*)canv->GetListOfPrimitives()->FindObject(padName.data());
+    pad->cd();
+    TH2F* h = (TH2F*)pad->GetListOfPrimitives()->FindObject(histName.data());
+    std::string titleh = h->GetTitle();
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:113]:
```suggestion
    const std::string titleh = h->GetTitle();
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    std::vector<std::string>::iterator it = std::find(badSectorsName.begin(), badSectorsName.end(), titleh);
+    if (it != badSectorsName.end()) {
+      TPaveText* msg = new TPaveText(0.1, 0.9, 0.9, 0.95, "NDC");
+      msg->SetBorderSize(1);
+      int index = std::distance(badSectorsName.begin(), it);
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:118]:
```suggestion
      const int index = std::distance(badSectorsName.begin(), it);
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  auto* canv = dynamic_cast<TCanvas*>(mo->getObject());
+  // Loop over all histograms and beautify accordingly to specific quality object
+  for (int tpads = 1; tpads <= 72; tpads++) {
+    auto padName = fmt::format("c_ROCs_Pedestal_2D_{:d}", tpads);
+    auto histName = fmt::format("h_Pedestals_ROC_{:02d}", tpads - 1);
+    TPad* pad = (TPad*)canv->GetListOfPrimitives()->FindObject(padName.data());
+    pad->cd();
+    TH2F* h = (TH2F*)pad->GetListOfPrimitives()->FindObject(histName.data());
+    std::string titleh = h->GetTitle();
+    std::vector<std::string>::iterator it = std::find(badSectorsName.begin(), badSectorsName.end(), titleh);
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:114]:
```suggestion
    auto it = std::find(badSectorsName.begin(), badSectorsName.end(), titleh);
```

File: Modules/TPC/src/PadCalibrationCheck2D.cxx [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    if (it != badSectorsName.end()) {
```
Comment (wiechula) [inline Modules/TPC/src/PadCalibrationCheck2D.cxx:115]: Change logic to make clear that the reader does not have to continue reading the code:
```suggestion
    if (it == badSectorsName.end()) {
      continue
    }
```

**Review by wiechula — CHANGES_REQUESTED** [review]
@lauraser , thanks for the code.
I have many small comments to improve code readability mainly. For the future, in general, please always
* use `const` whenever possible
* always use readability braces

**Review by wiechula — APPROVED** [review]

**Review by wiechula — APPROVED** [review]

**Comment by wiechula** [comment]
@Barthelemy , seem it is not merged.

**Pull request merged**